arangod/Transaction/IndexesSnapshot.cpp

```diff
@@ -0,0 +1,64 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2022 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Jan Steemann
+////////////////////////////////////////////////////////////////////////////////
+
+#include "IndexesSnapshot.h"
+
+#include "Indexes/Index.h"
+#include "StorageEngine/PhysicalCollection.h"
+
+namespace arangodb {
+
+IndexesSnapshot::IndexesSnapshot(PhysicalCollection& collection)
+    : _locker(collection._indexesLock, collection._indexesLockWriteOwner,
+              __FILE__, __LINE__),
+      _valid(true) {
+  _indexes.reserve(collection._indexes.size());
+  for (auto const& idx : collection._indexes) {
+    _indexes.emplace_back(idx);
+  }
```

Comment:
Why do we copy out the indexes? We only allow access to our index list as long as we have the locker, but if we have the locker, couldn't we just as well access to collection index list?

Reply:
In there we could access the `_indexes` property from `PhysicalCollection` directly, without having 2 copy anything.
There are two reasons why I would still prefer creating a copy:
1. `PhysicalCollection::_indexes` is of type `IndexesContainerType`, which in an `std::set` with a custom comparator. Some operations, such as `_indexes[0]` don't feel natural when performed on an `std::set`. I will probably also make more use of the IndexesSnapshot in the future, and will then potentially access members from position 1 on etc. All this is possible with an `std::set` somehow as well, but it feels awkward.
2. If I use a const reference to the original `_indexes` member of `PhysicalCollection` instead of creating a copy, it is still possible that someone gives up the lock (e.g. by calling `IndexesSnapshot::release()`) before accessing the member, which would be unsafe. Or acquire the reference, then call `release()` and after that use the reference. All these potential issues can be avoided by creating a copy.

Reply:
Ok, if we use a different container then it makes more sense.